### ISSUE TITLE
DT[TRUE] can lead to invalid key

### ISSUE THREAD

**Issue by mattdowle** [Rdatatable/data.table#3215]
I didn't realize `DT[TRUE]` was a way to achieve a shallow copy.  shallow copy was only intended for internal use. Thanks to @renkun-ken for highlighting this in #3214.

In v1.11.8 we see this :   
```R
DT = data.table(id = 1:5, key="id")
DT1 = DT[TRUE]
key(DT1)
[1] "id"
DT1[3, id:=6L]
key(DT1)
# NULL
DT$id
# [1] 1 2 6 4 5     # should be 1:5
key(DT)
# [1] "id"              # invalid key
```

It only occurs after `DT[TRUE]`, iiuc, which hopefully folk have not discovered or relied on too much?!   I hope the usage out there is like @renkun-ken described to add new columns to the shallow copy,  not to change existing columns!

**Comment by renkun-ken** [comment]
Yes, setkey, changing existing columns cannot be used on the shallow copy since columns themselves are not copied.

**Comment by jangorecki** [comment]
If we won't allow to make shallow copy with `dt[TRUE]` this issue will be automatically resolved.

**Comment by mattdowle** [comment]
Eventually. But in the meantime, we can't break @renkun-ken's workflow.
More detail here: https://github.com/Rdatatable/data.table/issues/3214#issuecomment-462490046

**Comment by jangorecki** [comment]
the following code could be added to tests to ensure copy behaviour
```r
DT = data.table(a=c(1,2), b=c("b","a"))
address(DT)
address(DT[])
address(DT[, .SD])
address(DT[TRUE])
sapply(DT, address)
sapply(DT[], address)
sapply(DT[, .SD], address)
sapply(DT[TRUE], address)
```

**Comment by MichaelChirico** [comment]
Now that we have regex search in GitHub we can get a better sense of how this is used downstream:

https://github.com/search?q=lang%3AR+%2F%28%5B%5E%21%3D%3E%3C%5D%3D%7C%3C-%29%5B+%5D*%5Ba-z0-9._%5D%2B%5C%5B%5B+%5D*TRUE%5B+%5D*%5C%5D%2F&type=code

(a bit hard to ensure it's `[.data.table` statically so this is all "assignment of `name[TRUE]`")

Really rare usage pattern indeed. Here's one usage that per the comment assumes `[TRUE]` is inducing a copy:

https://github.com/MiDataInt/mdi-apps-framework/blob/dd8124a1d7b00a36a84ae52165df9dc72ecf272b/shiny/shared/session/modules/appControl/historyList/historyList_server.R#L62
